### ISSUE TITLE
ArangoDB 2.5 Beta

### ISSUE THREAD

**Issue by CoDEmanX** [arangodb/arangodb#1259]
Installed Windows beta 2 64bit for all users to my usual folder, but it refuses to start. It seems like it expects a conf file to be in Program Files (x86), even if it wasn't installed there (but it is concatenated with cwd):

    C:\Users\Public\Documents\ArangoDB\bin>arangod
    2015-03-02T09:00:34Z [13736] INFO no system init file 'C:\Users\Public\Documents\ArangoDB\C:\Program Files (x86)\arangodb\etc\arangodbarangod.conf' found
    2015-03-02T09:00:34Z [13736] INFO please use the '--database.directory' option
    2015-03-02T09:00:34Z [13736] FATAL no database path has been supplied, giving up

**Comment by fceller** [comment]
That looks like a bug. Will check.

**Comment by CoDEmanX** [comment]
It also refuses to start if some other options aren't specified. It asks me to provide `--server.endpoint`, `--javascript.startup-directory` and `--javascript.app-path`.

If I supply relative paths for the parameters it asks for, a JS exception will be raised (possibly unrelated):

    2015-03-02T09:18:17Z [13452] ERROR stacktrace: Error: cannot use '..' to escape top-level-directory, prefix = '', path = '../share/arangodb/js/node/buffer.js/..

It works with absolute paths after a database upgrade:

    C:\Users\Public\Documents\ArangoDB\bin>arangod --database.directory "C:\Users\Public\Documents\ArangoDB\var\lib\arangodb" --server.endpoint tcp://0.0.0.0 --javascript.startup-directory "C:\Users\Public\Documents\ArangoDB\share\arangodb\js" --javascript.app-path "C:\Users\Public\Documents\ArangoDB\share\arangodb\js\apps"

**Comment by CoDEmanX** [comment]
In Aardvark, there seems to be an issue with permissions:

![arangodb_2 5_dashboard](https://cloud.githubusercontent.com/assets/1309249/6438219/6c5464f8-c0c7-11e4-99cd-f19da10b3ba8.png)

I can see the dashboard just fine for `_system`, but if a switch to another database, it won't let me see it or change database again (see the broken dropdown in above screenshot).

**Comment by CoDEmanX** [comment]
If I try to generate a Foxx app, it fails because it can't find `repository.js.tmpl`:

    Error: 2. File exists: while reading c:\Users\Public\Documents\ArangoDB\share\arangodb\js\server\modules\org\arangodb\foxx\templates\repository.js.tmpl: system error

The actual path is

    C:\Users\Public\Documents\ArangoDB\share\arangodb\js\apps\system\aardvark\templates\repository.js.tmpl

**Comment by mchacki** [comment]
Hi @CoDEmanX the behaviour of the web-interface is intended.
Only the _system database is collecting the information required for the dashboard.
So if your are using authentication and you have database _system with userA and database test with userB where userB has no access to _system it is not possible for him to see the statistics at all.

Also switching Databases is restricted to Databases your current user has access to.
Your test user does not seem to have access except the Test database.

**Comment by mchacki** [comment]
Thanks a lot for all the feedback so far. *thumbs up*

We will work through all of it and fix it.
Good work!

**Comment by CoDEmanX** [comment]
Ok, I see. But I disabled authentication altogether, shouldn't every user see everything? (Is there even a user if auth is off?)

**Comment by mchacki** [comment]
You are totally right. If auth is off, every user should see everything.
I can confirm it is broken then will fix. Thanks a lot.

**Comment by clintwood** [comment]
I aborted 2.5 Beta for same reasons above...

BTW, I tried `--configuration` switch to get it to look at the correct `arangod.conf` then got the permissions issues in web UI (as above)!

**Comment by dothebart** [comment]
2.5 Beta 3 has the directory issue fixed - installing it in Random locations should work again.

**Comment by clintwood** [comment]
Great! Thank you! Will give it a try!

**Comment by clintwood** [comment]
On 2.5 Beta 3, I can confirm that the path issue is resolved but the permissions issue as mentioned by @CoDEmanX still persists.

**Comment by dothebart** [comment]
@clintwood wfm on 8.1-server (service and ctrl+c in the console), can you give some more details?

**Comment by clintwood** [comment]
@dothebart, for all tests it was a clean install of 2.5 Beta with only _system DB (no user created DBs).

For the shutdown issue, Ctrl+c in a console window running as Administrator, I tested on Windows Server 2012 and there was no issue. In this case shutdown took about 8 seconds (which is longer than it used to but i don't think an issue).

However on Windows 8.1 (high spec machine i7 etc.) in Admin console window the shutdown takes around 80 seconds.  Also clean install no user DBs.
So then running as a service this is why the service control manager complains about not shutting down in time.  It does eventually shutdown but only after the SCM has complained.

One last thing that caused a lot of confusion was the log file is either not being truncated or not being appended too.  It appears that the logging code is just opening the file and writing from the beginning of the file irrespective of if there is anything in the file.  So if there are errors which make the file longer than normal startup logging takes, any extra lines beyond a successful startup log output remain in the file and any new logging just overwrites the beginning of the file leaving a wierd mixed up log file.  This should probably be a separate issue.

**Comment by clintwood** [comment]
@dothebart, the permissions issue as mentioned by @CoDEmanX still persists on both Windows 8.1 and Windows server 2012 that I used in the previous post.

**Comment by dothebart** [comment]
we started the doing beta3 release on friday; so the path issue was already identified and fixed the time you reported it. @mchacki is working on it.
Regarding the shutdown, do you see whats it actually doing in that time? cpu usage? disk io?
The logfile issue has to be researched; it should be appended, maybe another windows api weirdness..

**Comment by clintwood** [comment]
@dothebart, sorry for the delay, had to reinstall 2.5 Beta 3...

No cpu usage, and no disk IO - looks idle to me from process explorer.
I set the log level to `debug` and here is where the delay is in the logs:
```
2015-03-03T12:37:22Z [2724] DEBUG [..\..\lib\ApplicationServer\ApplicationServer.cpp:618] shutting down server feature 'V8'
2015-03-03T12:38:40Z [2724] DEBUG [..\..\lib\ApplicationServer\ApplicationServer.cpp:618] shutting down server feature 'scheduler'
```
FYI: `version 2.5.0-beta3 [win64]` on Windows 8.1

**Comment by pluma** [comment]
I've fixed the problem with the authentication error when authentication is disabled in e8f3ab9. But the logic in the frontend for determining which databases a user may access is still broken.

**Comment by Linicks** [comment]
I'm also seeing the permissions bug with 2.5 Beta3 on Ubuntu 14.10 x64.   When I create a database, I use the default root user with no password.  It creates the database, but doesn't activate it.  If I click on it, I get the permissions error, and the manage DB drop-down breaks.

**Comment by fceller** [comment]
Beta 4 is available for testing:

https://www.arangodb.com/download-beta

**Comment by CoDEmanX** [comment]
Default paths work again in beta 4 as far as I can tell. I can also create a new Foxx app and the interactive documentation operates as expected. There's an error however if I try to download it:

    {"error":"[ArangoError 3009: App not found at /foxx-test-1]"}

The mount point is `/foxx-test-1` in database `test `, the database parameter is missing in the URL however:

    http://localhost:8529/_admin/aardvark/foxxes/download/zip?mount=%Ffoxx-test-1

It works if I insert `/_db/test` after the port.

**Comment by jimac** [comment]
Installed windows beta 4 as a service in Windows 7.

* Web interface OK.
* Add Application from web interface generates only the new paths, the manifest, the setup and teardown scripts. Nothing else, though I entered a collection to create (not existent at that time). Shouldn't create any files in controllers, models & repositories dirs?
* I then tried to use foxx-manager. It prompt for a password, then FATAL no 'javascript.startup-directory' has been supplied, giving up. Checked conf file and is OK. 
* Same error trying arangosh.
* Starting foxx-manager with --javascript.startup-directory "C:\absolute path to javascript dir" worked  ....well almost, since instead of foxx-manager I just got an arango shell !!.

**Comment by fceller** [comment]
I've created a new installer package, @jimac can you please check if this fixes the problems?

https://www.arangodb.com/repositories/Windows7/x86_64/ArangoDB-2.5.0-beta4b-win64.exe

Thanks

**Comment by jimac** [comment]
@fceller nice done. The new installer for windows 7 seems that fixes all the before mentioned  problems.

Thank you,

**Comment by clintwood** [comment]
@fceller, on my system with `version 2.5.0-beta4b [win64]` on Windows 8.1 I am still experiencing the 80 second delay before shutdown as [mentioned here](https://github.com/arangodb/arangodb/issues/1259#issuecomment-76885417) and [here](https://github.com/arangodb/arangodb/issues/1259#issuecomment-76939115).

**Comment by dothebart** [comment]
@clintwood I've added some more debug log lines, could you retry and post them again?
I still can't reproduce this... 
Since it seems to be the garbage collection of V8 we're suspecting something still to be active; do you have a foxx app installed, or a browser window opened with the admin interface?

**Comment by clintwood** [comment]
@dothebart, will try now. No, clean install, literally run either as a service in which case SCM complains about service not shutting down in time or install into a folder and from command line run `arangod.exe` no changed settings.

**Comment by clintwood** [comment]
@dothebart, here are the offending entries in the log output:
```
2015-03-09T15:31:03Z [7548] DEBUG [C:\cygwin64\home\fceller\ArangoDB-2.5\arangod\V8Server\ApplicationV8.cpp:1149] Waiting for GC Thread to finish action
2015-03-09T15:31:03Z [7548] DEBUG [C:\cygwin64\home\fceller\ArangoDB-2.5\arangod\V8Server\ApplicationV8.cpp:1155] Commanding GC Thread to terminate
2015-03-09T15:32:18Z [7548] DEBUG [C:\cygwin64\home\fceller\ArangoDB-2.5\arangod\V8Server\ApplicationV8.cpp:1175] Shutting down V8
```
I can send you the complete log but I'm not sure that is required (if it let me know) as there are no errors just DEBUG & INFO entries.

**Comment by sleepycat** [comment]
The [download page](https://www.arangodb.com/download-beta) has a section on installing 2.5b4 on Ubuntu 14.04. Unfortunately 2.5 depends on libstdc++6 4.9 which is only available in 14.10+. I have backports enabled but it seems it is not in there: 

```
mike@sleepycat:~☺  sudo dpkg -iD 1 Downloads/arangodb_2.5.0b4_amd64.deb 
D000001: ensure_diversions: new, (re)loading
D000001: ensure_statoverrides: new, (re)loading
(Reading database ... 282268 files and directories currently installed.)
Preparing to unpack .../arangodb_2.5.0b4_amd64.deb ...
D000001: process_archive oldversionstatus=unpacked but not configured
D000001: ensure_diversions: same, skipping
Unpacking arangodb (2.5.0b4) over (2.5.0b4) ...
D000001: ensure_diversions: same, skipping
D000001: process_archive updating info directory
D000001: generating infodb hashfile
dpkg: dependency problems prevent configuration of arangodb:
 arangodb depends on libstdc++6 (>= 4.9); however:
  Version of libstdc++6:amd64 on system is 4.8.2-19ubuntu1.

dpkg: error processing package arangodb (--install):
 dependency problems - leaving unconfigured
Processing triggers for ureadahead (0.100.0-16) ...
D000001: ensure_diversions: same, skipping
Processing triggers for man-db (2.6.7.1-1ubuntu1) ...
D000001: ensure_diversions: same, skipping
Errors were encountered while processing:
 arangodb
```
The libstdc++ versions for each Ubuntu release are listed [here](http://packages.ubuntu.com/search?keywords=libstdc%2B%2B6).

**Comment by fceller** [comment]
@sleepycat can you check if you downloaded the 14.10 by accident? I've tried the 2.5.0 and it only happened when I tried to install the 14.10 version.

**Comment by sleepycat** [comment]
I'm was running 14.04 and was pulling from the repo at https://www.arangodb.com/repositories/arangodb2/xUbuntu_14.04
I'm pretty sure all I did was replace arangodb2 with unstable in that URL and then update/upgrade and that's when things went pear-shaped.

I ended up upgrading to 14.10 and everything is working nicely.

**Comment by fceller** [comment]
@sleepycat good to hear. Will close the ticket.
